### ISSUE TITLE
Scopes proposal

### ISSUE THREAD

**Issue by orsenkucher** [felangel/hydrated_bloc#54]
### Preview
**Protoversion** of this change is available on my [scopes-implicit](https://github.com/orsenkucher/hydrated_bloc/tree/scopes-implicit)

### Resolved issues
* Obligation to modify constructor is now completely relaxed
* `HydratedProvider` does not depend on `hydrated_bloc` now, so can be merged with `BlocProvider`

### Issues
* Interference with `package:hive`? (`Hive.init` in multiple places)
* Provisioning of `directory` should be done only once/or redone after all boxes are closed
* Create scopes dynamically, i.e. encrypted scope after user logged in

### Preview
**Protoversion** of this change is available on my [scopes](https://github.com/orsenkucher/hydrated_bloc/tree/scopes) branch

### Usage
Inside `main()` we configure `Hydrated` to init some storage boxes with tokens `'secure'` and `'number'`:
```dart
void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  BlocSupervisor.delegate = await HydratedBlocDelegate.build();

  final byteskey = sha256.convert(utf8.encode('password')).bytes;
  await Hydrated.config({
    'secure': ScopeConfig(encryptionCipher: HydratedAesCipher(byteskey)),
    'number': ScopeConfig(),
  });

  runApp(App());
}
```

Inside `build()` then, we use `Hydrated.scope()` to obtain `ScopeBuilder` and then build `HydratedScope` with `child` widget, for example `HydratedProvider`:
```dart
Widget build(BuildContext context) {
  return Hydrated.scope('number')(HydratedProvider<CounterBloc>(
    create: (context, scope) => CounterBloc(scope),
    child: Builder(builder: (context) {
      return Hydrated.scope('secure')(HydratedProvider<CounterBloc>(
        create: (context, scope) => CounterBloc(scope),
        child: MaterialApp(
          title: 'Flutter Demo',
          home: CounterPage(context.bloc<CounterBloc>()),
        ),
      ));
    }),
  ));
}
```

As you can see, 
```dart
create: (context, scope) => CounterBloc(scope) 
```
now has second parameter - `scope` (or `scopeToken`/`scopeName`) i.e `'secure'`,
which is extracted from context in `HydratedProvider`'s `build()` method.
It has to be passed down to `CounterBloc` via it's constructor:
```dart
class CounterBloc extends HydratedBloc<CounterEvent, CounterState> {
  CounterBloc(String scope) : super(scope);
...
```
So now user is responsible for providing `scopeToken` to `super`, as we can't pull data from `InheritedWidget` inside core lib.

And also **custom** `HydratedProvider` is not the most pleasant thing to have

**Comment by felangel** [comment]
closing this in favor of conversation in https://github.com/felangel/cubit/issues/67
